compiler/src/dotty/tools/backend/sjs/JSExportsGen.scala

```diff
@@ -0,0 +1,914 @@
+        js.LoadJSConstructor(encodeClassName(superClassSym))
+    }
+
+    val receiver = js.This()(jstpe.AnyType)
+    val nameTree = genExpr(sym.jsName)
+
+    if (sym.isJSGetter) {
+      assert(allArgs.isEmpty,
+          s"getter symbol $sym does not have a getter signature")
+      js.JSSuperSelect(superClass, receiver, nameTree)
+    } else if (sym.isJSSetter) {
+      assert(allArgs.size == 1 && allArgs.head.isInstanceOf[js.Tree],
+          s"setter symbol $sym does not have a setter signature")
+      js.Assign(js.JSSuperSelect(superClass, receiver, nameTree),
+          allArgs.head.asInstanceOf[js.Tree])
+    } else {
+      js.JSSuperMethodCall(superClass, receiver, nameTree, allArgs)
+    }
+  }
+
+  private def genApplyForSingleExportedNonJSSuperCall(
+      formalArgsRegistry: FormalArgsRegistry, exported: Exported, static: Boolean): js.Tree = {
+
+    implicit val pos = exported.pos
+
+    // the (single) type of the repeated parameter if any
+    val repeatedTpe = exported.params.lastOption.withFilter(_.isRepeated).map(_.info)
+
+    val normalArgc = exported.params.size - (if (repeatedTpe.isDefined) 1 else 0)
+
+    // optional repeated parameter list
+    val jsVarArgPrep = repeatedTpe map { tpe =>
+      val rhs = genJSArrayToVarArgs(formalArgsRegistry.genVarargRef(normalArgc))
+      val ident = freshLocalIdent("prep" + normalArgc)
+      js.VarDef(ident, NoOriginalName, rhs.tpe, mutable = false, rhs)
+    }
+
+    // normal arguments
+    val jsArgRefs =
+      (0 until normalArgc).toList.map(formalArgsRegistry.genArgRef(_))
+
+    // Generate JS code to prepare arguments (default getters and unboxes)
+    val jsArgPrep = genPrepareArgs(jsArgRefs, exported, static) ++ jsVarArgPrep
+    val jsArgPrepRefs = jsArgPrep.map(_.ref)
+
+    // Combine prep'ed formal arguments with captures
+    val allJSArgs = {
+      exported.captureParamsFront.map(_.ref) :::
+      jsArgPrepRefs :::
+      exported.captureParamsBack.map(_.ref)
+    }
+
+    val jsResult = genResult(exported, allJSArgs, static)
+
+    js.Block(jsArgPrep :+ jsResult)
+  }
+
+  /** Generate the necessary JavaScript code to prepare the arguments of an
+   *  exported method (unboxing and default parameter handling)
+   */
+  private def genPrepareArgs(jsArgs: List[js.Tree], exported: Exported, static: Boolean)(
+      implicit pos: SourcePosition): List[js.VarDef] = {
+
+    val result = new mutable.ListBuffer[js.VarDef]
+
+    for {
+      (jsArg, (param, i)) <- jsArgs.zip(exported.params.zipWithIndex)
+    } yield {
+      // Unboxed argument (if it is defined)
+      val unboxedArg = unbox(jsArg, param.info)
+
+      // If argument is undefined and there is a default getter, call it
+      val verifiedOrDefault = if (param.hasDefault) {
+        js.If(js.BinaryOp(js.BinaryOp.===, jsArg, js.Undefined()), {
+          genCallDefaultGetter(exported.sym, i, param.sym.sourcePos, static) {
+            prevArgsCount => result.take(prevArgsCount).toList.map(_.ref)
```

Comment:
I tried to do as you suggest, but it gets uglier than expected because we then also need to compute the `tpe`s in advance (not just the idents), and to compute the tpes we need to also look at whether it's repeated param or not (and if it is, it's a Scala `Seq` type, so it's annoying to produce), etc. In the end it's harder to reason about than the mutable state, IMO.